api/server/server.go

```diff
@@ -1091,6 +1091,179 @@ func postContainersCopy(eng *engine.Engine, version version.Version, w http.Resp
 	return nil
 }
 
+func handleCopyError(w http.ResponseWriter, statusp *int, errp *error) {
+	status, err := *statusp, *errp
+
+	if err == nil {
+		return
+	}
+
+	log.Errorf("%d: %s", status, err)
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(status)
+
+	errorMessage := map[string]string{
+		"detail": err.Error(),
+	}
+
+	encoder := json.NewEncoder(w)
+	encoder.Encode(errorMessage)
+}
+
+// getContainersCopy copies a file or directory from a container to the client.
+func getContainersCopy(eng *engine.Engine, version version.Version, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
+	var (
+		err        error
+		statusCode int
+		sourcePath string
+	)
+
+	defer handleCopyError(w, &statusCode, &err)
+
+	if vars == nil {
+		statusCode = http.StatusInternalServerError
+		err = fmt.Errorf("missing parameter")
+		return nil
```

Comment:
Related to the comment above, about using the shared httpError method, this feels a little too cute to return errors in this indirect way.  A quick read of the code will lead people to think the "return nil" means no error, when in fact the defer does magic and actually does return an error.  I think this will lead to confusion.